split.go

```diff
@@ -0,0 +1,114 @@
+package girc
+
+import (
+	"strconv"
+	"strings"
+)
+
+// splitFunc is the type used for functions implementing splitting of
+// too long IRC messages. The function is passed an event which must be
+// split into multiple and the maximum event length.
+type splitFunc func(event *Event, maxLen int) []*Event
+
+var splitFuncs = map[string]splitFunc{
+	PRIVMSG: splitPRIVMSG,
+}
+
+// maxHostLen returns the maximum possible length of a server message
+// prefix as defined by the following ABNF in RFC 2812:
+//
+//   [ ":" prefix SPACE ]
+//
+func maxPrefixLen(client *Client) int {
+	var nicklen, hostlen int
+	var err error
+
+	nickval, success := client.GetServerOption("MAXNICKLEN")
+	if success {
+		nicklen, err = strconv.Atoi(nickval)
+	}
+	if !success || err != nil {
+		nicklen = 10 // common value according to irssi
+	}
+
+	hostval, success := client.GetServerOption("HOSTLEN")
+	if success {
+		hostlen, err = strconv.Atoi(hostval)
+	}
+	if !success || err != nil {
+		hostlen = 63 // common value according to irssi
+	}
+
+	// From RFC 2812:
+	//
+	//   prefix = servername / ( nickname [ [ "!" user ] "@" host ] )
+	//
+	// The code here assumes the `servername` is never used as a
+	// prefix. As this function is concerned with messages send by a
+	// client. In accordance with the ABNF from above, the maximum
+	// length is therefore calculated as:
+	//
+	//   ":" <nick> "!" <host> " "
+	//
+	return 1 + nicklen + 1 + hostlen + 1
+}
+
+func splitPRIVMSG(event *Event, maxLen int) []*Event {
+	newMsg := func(text string) *Event {
+		e := event.Copy()
+		e.Params[len(event.Params)-1] = text
+		return e
+	}
+
+	// Calculate the maximum text message length (excluding targets).
+	rawEvent := event.Copy()
+	rawEvent.Params = event.Params[0 : len(event.Params)-1]
+	maxTextLen := maxLen - rawEvent.Len() - 1 // -1 for potential :
```

Comment:
Use:

```Go
maxTextLen := maxLen - rawEvent.Len() - len(" :")
```

instead as colon **and** space for the trailing parameter are not included in the `rawEvent.len()` calculation?

Reply:
Went ahead with this, being too conservative shouldn't be an issue in regards to the maximum length. The only disadvantage is that we might send more splits than necessarily.